File: frontend/src/lib/autonomyApi.ts
import { apiUrl } from './backendBase';

async function request<T>(path: string, init?: RequestInit): Promise<T> {
    const res = await fetch(apiUrl(path), {
        ...init,
        headers: {
            'Content-Type': 'application/json',
            'X-JimAI-CSRF': '1',
            ...(init?.headers || {}),
        },
    });
    if (!res.ok) {
        const text = await res.text().catch(() => '');
        throw new Error(`${res.status} ${res.statusText} ${text}`.trim());
    }
    if (res.status === 204) return null as T;
    return (await res.json()) as T;
}

export type EpisodeRecord = {
    id: string;
    timestamp: number;
    run_id: string;
    agent_id: string;
    event: string;
    outcome: string;
    summary: string;
    metadata?: Record<string, unknown>;
};

export type SkillEntry = {
    id: string;
    name: string;
    description: string;
    objective: string;
    artifact_type: string;
    artifact: string;
    tags: string[];
    success_count: number;
    use_count: number;
    last_used_at: number;
    verifier_score: number;
    created_at: number;
    updated_at: number;
};

export type ReflectionTrace = {
    id: string;
    created_at: number;
    run_id: string;
    agent_id: string;
    objective: string;
    attempt: number;
    failure_reason: string;
    lesson: string;
};

export type HeartbeatJob = {
    id: string;
    name: string;
    objective: string;
    interval_seconds: number;
    next_fire_at: number;
    last_fired_at: number;
    fire_count: number;
    enabled: boolean;
    one_shot: boolean;
    last_error: string;
};

export type HeartbeatStatus = {
    running: boolean;
    bound: boolean;
    job_count: number;
    enabled_count?: number;
    tick_interval_seconds?: number;
    last_tick_at?: number;
    last_tick_fired?: number;
    last_error?: string;
};

export const memoryStats = () => request<Record<string, number>>('/api/autonomy/memory/stats');
export const memoryRecent = (limit = 50) =>
    request<{ items: EpisodeRecord[] }>(`/api/autonomy/memory/recent?limit=${limit}`);
export const memorySearch = (query: string, limit = 5) =>
    request<{ results: (EpisodeRecord & { score: number })[] }>('/api/autonomy/memory/search', {
        method: 'POST',
        body: JSON.stringify({ query, limit }),
    });
export const memoryConsolidate = (max_per_run = 3) =>
    request<{ consolidated: number }>('/api/autonomy/memory/consolidate', {
        method: 'POST',
        body: JSON.stringify({ max_per_run }),
    });

export const skillStats = () => request<Record<string, unknown>>('/api/autonomy/skills/stats');
export const skillList = (limit = 200) =>
    request<{ items: SkillEntry[] }>(`/api/autonomy/skills?limit=${limit}`);
export const skillRetrieve = (objective: string, limit = 5) =>
    request<{ results: (SkillEntry & { score: number })[] }>('/api/autonomy/skills/retrieve', {
        method: 'POST',
        body: JSON.stringify({ objective, limit }),
    });
export const skillDelete = (id: string) =>
    request<{ ok: boolean }>(`/api/autonomy/skills/${encodeURIComponent(id)}`, { method: 'DELETE' });

export const reflectionStats = () => request<Record<string, unknown>>('/api/autonomy/reflections/stats');
export const reflectionLookup = (objective: string, limit = 4) =>
    request<{ lessons: ReflectionTrace[] }>('/api/autonomy/reflections/lookup', {
        method: 'POST',
        body: JSON.stringify({ objective, limit }),
    });

export const heartbeatStatus = () => request<HeartbeatStatus>('/api/autonomy/heartbeat/status');
export const heartbeatJobs = () =>
    request<{ items: HeartbeatJob[] }>('/api/autonomy/heartbeat/jobs');
export const heartbeatStart = () => request<HeartbeatStatus>('/api/autonomy/heartbeat/start', { method: 'POST' });
export const heartbeatStop = () => request<HeartbeatStatus>('/api/autonomy/heartbeat/stop', { method: 'POST' });
export const heartbeatTick = () =>
    request<{ due: number; fired: number; errors: unknown[] }>('/api/autonomy/heartbeat/tick', { method: 'POST' });
export const heartbeatAddJob = (input: {
    name: string;
    objective: string;
    interval_seconds?: number;
    one_shot?: boolean;
    enabled?: boolean;
    first_fire_in_seconds?: number;
}) =>
    request<HeartbeatJob>('/api/autonomy/heartbeat/jobs', {
        method: 'POST',
        body: JSON.stringify(input),
    });
export const heartbeatDeleteJob = (id: string) =>
    request<{ ok: boolean }>(`/api/autonomy/heartbeat/jobs/${encodeURIComponent(id)}`, { method: 'DELETE' });

File: backend/api/autonomy_api.py
"""Autonomy primitives API.

Exposes episodic memory, skill library, reflections, replan evaluator, and
the heartbeat scheduler as a small REST surface. Read-mostly; mutations are
gated by the existing CSRF + API-key middleware.
"""

from __future__ import annotations

import logging
from typing import Any

from fastapi import APIRouter, HTTPException

from agent_space.autonomy import runtime as autonomy_runtime

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/autonomy", tags=["autonomy"])


# --------------------------------------------------------------------- memory


@router.get("/memory/stats")
async def memory_stats() -> dict[str, Any]:
    return autonomy_runtime.get_episodic_memory().stats()


@router.get("/memory/recent")
async def memory_recent(limit: int = 50) -> dict[str, Any]:
    rows = autonomy_runtime.get_episodic_memory().list_recent(limit=limit)
    return {"items": [r.to_dict() for r in rows]}


@router.get("/memory/run/{run_id}")
async def memory_for_run(run_id: str, limit: int = 200) -> dict[str, Any]:
    rows = autonomy_runtime.get_episodic_memory().list_by_run(run_id, limit=limit)
    return {"run_id": run_id, "items": [r.to_dict() for r in rows]}


@router.post("/memory/search")
async def memory_search(payload: dict[str, Any]) -> dict[str, Any]:
    query = str(payload.get("query") or "").strip()
    if not query:
        raise HTTPException(status_code=400, detail="query is required")
    limit = int(payload.get("limit") or 5)
    min_score = float(payload.get("min_score") or 0.55)
    hits = await autonomy_runtime.get_episodic_memory().search(
        query, limit=limit, min_score=min_score
    )
    return {
        "query": query,
        "results": [
            {**rec.to_dict(), "score": round(score, 4)}
            for rec, score in hits
        ],
    }


@router.post("/memory/consolidate")
async def memory_consolidate(payload: dict[str, Any] | None = None) -> dict[str, Any]:
    payload = payload or {}
    max_per_run = int(payload.get("max_per_run") or 3)
    consolidated = await autonomy_runtime.get_episodic_memory().consolidate(
        max_per_run=max_per_run
    )
    return {"consolidated": consolidated}


# ---------------------------------------------------------------- skill library


@router.get("/skills/stats")
async def skills_stats() -> dict[str, Any]:
    return autonomy_runtime.get_skill_library().stats()


@router.get("/skills")
async def skills_list(limit: int = 200) -> dict[str, Any]:
    rows = autonomy_runtime.get_skill_library().list_all(limit=limit)
    return {"items": [r.to_dict() for r in rows]}


@router.post("/skills/retrieve")
async def skills_retrieve(payload: dict[str, Any]) -> dict[str, Any]:
    objective = str(payload.get("objective") or "").strip()
    if not objective:
        raise HTTPException(status_code=400, detail="objective is required")
    limit = int(payload.get("limit") or 5)
    artifact_type = payload.get("artifact_type")
    hits = await autonomy_runtime.get_skill_library().retrieve(
        objective,
        limit=limit,
        artifact_type=str(artifact_type) if artifact_type else None,
    )
    return {
        "objective": objective,
        "results": [
            {**entry.to_dict(), "score": round(score, 4)}
            for entry, score in hits
        ],
    }


@router.post("/skills/capture")
async def skills_capture(payload: dict[str, Any]) -> dict[str, Any]:
    name = str(payload.get("name") or "").strip()
    if not name:
        raise HTTPException(status_code=400, detail="name is required")
    artifact = str(payload.get("artifact") or "").strip()
    if not artifact:
        raise HTTPException(status_code=400, detail="artifact is required")
    entry = await autonomy_runtime.get_skill_library().capture(
        name=name,
        description=str(payload.get("description") or ""),
        objective=str(payload.get("objective") or ""),
        artifact=artifact,
        artifact_type=str(payload.get("artifact_type") or "code"),
        tags=list(payload.get("tags") or []),
        verifier_score=float(payload.get("verifier_score") or 1.0),
        metadata=dict(payload.get("metadata") or {}),
    )
    return entry.to_dict()


@router.delete("/skills/{entry_id}")
async def skills_delete(entry_id: str) -> dict[str, Any]:
    ok = await autonomy_runtime.get_skill_library().delete(entry_id)
    if not ok:
        raise HTTPException(status_code=404, detail="skill not found")
    return {"ok": True, "id": entry_id}


# ------------------------------------------------------------------ reflection


@router.get("/reflections/stats")
async def reflections_stats() -> dict[str, Any]:
    return autonomy_runtime.get_reflection_engine().stats()


@router.post("/reflections/lookup")
async def reflections_lookup(payload: dict[str, Any]) -> dict[str, Any]:
    objective = str(payload.get("objective") or "").strip()
    if not objective:
        raise HTTPException(status_code=400, detail="objective is required")
    limit = int(payload.get("limit") or 4)
    rows = autonomy_runtime.get_reflection_engine().lessons_for(objective, limit=limit)
    return {
        "objective": objective,
        "lessons": [r.to_dict() for r in rows],
    }


# --------------------------------------------------------------------- heartbeat


@router.get("/heartbeat/status")
async def heartbeat_status() -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        return {"running": False, "bound": False, "job_count": 0}
    status = sched.status()
    status["bound"] = True
    return status


@router.get("/heartbeat/jobs")
async def heartbeat_jobs() -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        return {"items": []}
    rows = sched.list_jobs()
    return {"items": [r.to_dict() for r in rows]}


@router.post("/heartbeat/start")
async def heartbeat_start() -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    return await sched.start()


@router.post("/heartbeat/stop")
async def heartbeat_stop() -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    return await sched.stop()


@router.post("/heartbeat/tick")
async def heartbeat_tick() -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    return await sched.tick()


@router.post("/heartbeat/jobs")
async def heartbeat_add_job(payload: dict[str, Any]) -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    name = str(payload.get("name") or "").strip()
    objective = str(payload.get("objective") or "").strip()
    if not name or not objective:
        raise HTTPException(status_code=400, detail="name and objective are required")
    job = await sched.add_job(
        name=name,
        objective=objective,
        interval_seconds=float(payload.get("interval_seconds") or 900.0),
        one_shot=bool(payload.get("one_shot") or False),
        first_fire_in_seconds=(
            float(payload["first_fire_in_seconds"])
            if payload.get("first_fire_in_seconds") is not None
            else None
        ),
        payload=dict(payload.get("payload") or {}),
        metadata=dict(payload.get("metadata") or {}),
        enabled=bool(payload.get("enabled", True)),
    )
    return job.to_dict()


@router.patch("/heartbeat/jobs/{job_id}")
async def heartbeat_update_job(job_id: str, payload: dict[str, Any]) -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    try:
        job = await sched.update_job(job_id, payload or {})
    except FileNotFoundError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc
    return job.to_dict()


@router.delete("/heartbeat/jobs/{job_id}")
async def heartbeat_delete_job(job_id: str) -> dict[str, Any]:
    sched = autonomy_runtime.get_heartbeat_scheduler()
    if sched is None:
        raise HTTPException(status_code=503, detail="heartbeat scheduler not bound")
    ok = await sched.delete_job(job_id)
    if not ok:
        raise HTTPException(status_code=404, detail="job not found")
    return {"ok": True, "id": job_id}
